lang/java/protobuf/src/main/java/org/apache/avro/protobuf/ProtoConversions.java

```diff
@@ -0,0 +1,141 @@
+import org.apache.avro.LogicalTypes;
+import org.apache.avro.Schema;
+
+public class ProtoConversions {
+
+  private static final int THOUSAND = 1000;
+  private static final int MILLION = 1000000;
+
+  // second value must be from 0001-01-01T00:00:00Z to 9999-12-31T23:59:59Z
+  // inclusive.
+  public static final long SECONDS_LOWERLIMIT = -62135596800L;
+  public static final long SECONDS_UPPERLIMIT = 253402300799L;
+
+  // nano value Must be from 0 to 999,999,999 inclusive.
+  public static final int NANOSECONDS_LOWERLIMIT = 0;
+  public static final int NANOSECONDS_UPPERLIMIT = 999999999;
+
+  // timestamp precise of conversion from long
+  public enum TimestampPrecise {
+    Millis, Micros
+  };
+
+  public static class TimestampMillisConversion extends Conversion<Timestamp> {
+    @Override
+    public Class<Timestamp> getConvertedType() {
+      return Timestamp.class;
+    }
+
+    @Override
+    public String getLogicalTypeName() {
+      return "timestamp-millis";
+    }
+
+    @Override
+    public Timestamp fromLong(Long millisFromEpoch, Schema schema, LogicalType type) throws IllegalArgumentException {
+      return ProtoConversions.fromLong(millisFromEpoch, TimestampPrecise.Millis);
+    }
+
+    @Override
+    public Long toLong(Timestamp value, Schema schema, LogicalType type) {
+      return ProtoConversions.toLong(value, TimestampPrecise.Millis);
+    }
+
+    @Override
+    public Schema getRecommendedSchema() {
+      return LogicalTypes.timestampMillis().addToSchema(Schema.create(Schema.Type.LONG));
+    }
+  }
+
+  public static class TimestampMicrosConversion extends Conversion<Timestamp> {
+    @Override
+    public Class<Timestamp> getConvertedType() {
+      return Timestamp.class;
+    }
+
+    @Override
+    public String getLogicalTypeName() {
+      return "timestamp-micros";
+    }
+
+    @Override
+    public Timestamp fromLong(Long microsFromEpoch, Schema schema, LogicalType type) throws IllegalArgumentException {
+      return ProtoConversions.fromLong(microsFromEpoch, TimestampPrecise.Micros);
+    }
+
+    @Override
+    public Long toLong(Timestamp value, Schema schema, LogicalType type) {
+      return ProtoConversions.toLong(value, TimestampPrecise.Micros);
+    }
+
+    @Override
+    public Schema getRecommendedSchema() {
+      return LogicalTypes.timestampMicros().addToSchema(Schema.create(Schema.Type.LONG));
+    }
+  }
+
+  private static long toLong(Timestamp value, TimestampPrecise precise) {
+    long rv = 0L;
+
+    switch (precise) {
+    case Millis:
+      rv = value.getSeconds() * THOUSAND + value.getNanos() / MILLION;
+      break;
+    case Micros:
+      rv = value.getSeconds() * MILLION + value.getNanos() / THOUSAND;
+      break;
+    }
+
+    return rv;
+  }
+
+  private static Timestamp fromLong(Long epoch, TimestampPrecise precise) throws IllegalArgumentException {
+    long seconds = 0L;
+    int nanos = 0;
+
+    switch (precise) {
+    case Millis:
+      seconds = epoch / THOUSAND;
```

Comment:
This will produce incorrect results for timestamps before the epoch, because [integer division in Java always rounds toward 0](https://docs.oracle.com/javase/specs/jls/se11/html/jls-15.html#jls-15.17.2). This means that this code will split -1.2 seconds into -1 whole seconds and -.2 split seconds, instead of turning into -2 whole seconds and +0.8 split seconds as would be required by the [rules of the protobuf Timestamp type](https://github.com/protocolbuffers/protobuf/blob/825c5bf4e9bfb70374da7a9fca4ad9a4997d44f1/src/google/protobuf/timestamp.proto#L133).

To make this properly work for timestamps before the epoch, you should always [round towards negative infinity](https://stackoverflow.com/questions/10457208/).

Reply:
Thanks the comment, I was missing that ...